base/serialize.jl

```diff
@@ -14,73 +14,98 @@ abstract LongTuple
+    :(=), :null, :gotoifnot, :A, :B, :C, :M, :N, :T, :S, :X, :Y,
+    :a, :b, :c, :d, :e, :f, :g, :h, :i, :j, :k, :l, :m, :n, :o,
+    :p, :q, :r, :s, :t, :u, :v, :w, :x, :y, :z,
+    :add_int, :sub_int, :mul_int, :add_float, :sub_float,
+    :mul_float, :unbox, :box,
+    :eq_int, :slt_int, :sle_int, :ne_int,
+    :arrayset, :arrayref,
+    :Core, :Base, svec(), :reserved16,
+    :reserved17, :reserved18, :reserved19, :reserved20,
+    false, true, nothing, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
+    12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27,
+    28, 29, 30, 31, 32
+]
+
 const ser_version = 2 # do not make changes without bumping the version #!
-const ser_tag = ObjectIdDict()
-const deser_tag = ObjectIdDict()
+const SER_TAG = ObjectIdDict()
+const DESER_TAG = ObjectIdDict()
 let i = 2
-    global ser_tag, deser_tag
-    for t = Any[
-             Symbol, Int8, UInt8, Int16, UInt16, Int32, UInt32,
-             Int64, UInt64, Int128, UInt128, Float32, Float64, Char, Ptr,
-             DataType, UnionType, Function,
-             Tuple, Array, Expr, LongSymbol, LongTuple, LongExpr,
-             LineNumberNode, SymbolNode, LabelNode, GotoNode,
-             QuoteNode, TopNode, TypeVar, Box, LambdaStaticData,
-             Module, UndefRefTag, Task, ASCIIString, UTF8String,
-             UTF16String, UTF32String, Float16,
-             SimpleVector, :reserved10, :reserved11, :reserved12,
-
-             (), Bool, Any, :Any, Bottom, :reserved21, :reserved22, Type,
-             :Array, :TypeVar, :Box,
-             :lambda, :body, :return, :call, symbol("::"),
-             :(=), :null, :gotoifnot, :A, :B, :C, :M, :N, :T, :S, :X, :Y,
-             :a, :b, :c, :d, :e, :f, :g, :h, :i, :j, :k, :l, :m, :n, :o,
-             :p, :q, :r, :s, :t, :u, :v, :w, :x, :y, :z,
-             :add_int, :sub_int, :mul_int, :add_float, :sub_float,
-             :mul_float, :unbox, :box,
-             :eq_int, :slt_int, :sle_int, :ne_int,
-             :arrayset, :arrayref,
-             :Core, :Base, svec(), :reserved16,
-             :reserved17, :reserved18, :reserved19, :reserved20,
-             false, true, nothing, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
-             12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27,
-             28, 29, 30, 31, 32]
-        ser_tag[t] = Int32(i)
-        deser_tag[Int32(i)] = t
+    for t = TAGS
+        SER_TAG[t] = Int32(i)
+        DESER_TAG[Int32(i)] = t
         i += 1
     end
 end
 
-# tags >= this just represent themselves, their whole representation is 1 byte
-const VALUE_TAGS = ser_tag[()]
-
-const EMPTY_TUPLE_TAG = ser_tag[()]
-const ZERO_TAG = ser_tag[0]
-const INT_TAG = ser_tag[Int]
+const NTAGS = length(TAGS)
+const TAG_PTRS = Array(Ptr{Void}, NTAGS)
 
-writetag(s, x) = write(s, UInt8(ser_tag[x]))
+function sertag(v::ANY)
+    idx = 2
+    ptr = pointer_from_objref(v)
+    @inbounds @simd for i = 1:NTAGS
+        ptr == TAG_PTRS[i] && return Int32(idx)
```

Comment:
I suspect it would be just as fast to do `ptr == pointer_from_objref(TAGS[i])`. `pointer_from_objref` is a no-op.